### PR TITLE
[CB] Refactor: extract common scheduling logic into Scheduler base class

### PR DESCRIPTION
**Pull request by dstekanov**
# What does this PR do?

This PR refactors the FIFOScheduler and PrefillFirstScheduler classes to extract common scheduling logic into the base Scheduler class (fixed TODO: further common-ize the two classes)

#### Local verification
Verified basic continuous batching utilities locally on CPU (macOS) by running existing tests:

```bash
pytest -q tests/generation/test_continuous_batching.py -k group_layers
pytest -q tests/generation/test_continuous_batching.py -k attention_mask
```

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? 
- [ ] Did you write any new necessary tests?

## Who can review?

@remi-or @ArthurZucker @McPatate

### CODE REVIEW

**Comment by remi-or** [comment]
This is great, and the change was long overdue tbh. But we just merged a PR that moved quite a bit of code in the schedulers, so I think you need to rebase + resolve conflicts with care. Sorry for the untimeliness of this all!

File: src/transformers/generation/continuous_batching/scheduler.py [hunk 1]
```diff
@@ -329,62 +355,16 @@ def schedule_batch(self, token_budget: int, cache_budget: int) -> list[RequestSt
...
-
-            # Early exit of the loop if we have no budget left
-            if token_budget == 0 or cache_budget == 0:
-                break
+        scheduled_requests, one_allocation_failed = self._process_candidates(
+            candidates,
+            token_budget,
+            cache_budget,
+            request_ids_to_remove_from_waiting,
+            safety_margin=None,
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/scheduler.py]: If we set safety_margins to 0 as a default value, then we are never outside of the margins: 0 would be better than None here

File: src/transformers/generation/continuous_batching/scheduler.py [hunk 2]
```diff
@@ -189,54 +189,34 @@ def _schedule_request(
             state.remaining_prefill_tokens = request_tokens[token_budget:]
             state.tokens_to_process = request_tokens[:token_budget]
 
+    def _process_candidates(
+        self,
+        candidates: list[RequestState],
+        token_budget: int,
+        cache_budget: int,
+        request_ids_to_remove_from_waiting: set[str],
+        safety_margin: float | None = None,
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/scheduler.py]: 0 would be a better default (cf. later comment)

File: src/transformers/generation/continuous_batching/scheduler.py [hunk 3]
```diff
@@ -189,54 +189,34 @@ def _schedule_request(
...
         for state in candidates:
-            # If we are out the safety margin, we only accept decoding requests or the first prefill request
             num_free_blocks = self.cache.get_num_free_blocks()
-            outside_safety_margin = num_free_blocks < safety_margins
-            if outside_safety_margin and scheduled_requests and state.status != RequestStatus.DECODING:
-                logger.info(
-                    f"Outside safety margin, breaking out of scheduling loop. {num_free_blocks = } {safety_margins = }"
-                )
-                break
+            if safety_margins is not None:
```
Comment (remi-or) [inline src/transformers/generation/continuous_batching/scheduler.py]: remove the check because `safety_margins` will never be None

**Review by remi-or — COMMENTED** [review]
LGTM, thanks for the contribution! If you can make the changes I commented about it would be awesome. Once that's done I will test, approve and merge :slightly_smiling_face:

**Comment by dstekanov** [comment]
> LGTM, thanks for the contribution! If you can make the changes I commented about it would be awesome. Once that's done I will test, approve and merge 🙂

Thanks! I’ve applied the requested changes

**Review by remi-or — APPROVED** [review]
LGTM! Thanks for the contribution, tests pass

**Comment by remi-or** [comment]
@bot /style

**Comment by remi-or** [comment]
It seems the style bot is broken, could you run `make style` please?

**Comment by dstekanov** [comment]
> It seems the style bot is broken, could you run `make style` please?

Just pushed a small style fix, seems I introduced it in my latest commit
